### ISSUE TITLE
The Gradle plugin generates invalid Dockerfile when choosing Java 21

### ISSUE THREAD

**Issue by saw303** [micronaut-projects/micronaut-starter#2281]
### Expected Behavior

If the Gradle build set Java version 21 like

```kotlin
java {
    sourceCompatibility = JavaVersion.toVersion("21")
    targetCompatibility = JavaVersion.toVersion("21")
}
```

the plugin should use the compatible base images for `Dockerfile` and `DockerfileNative`

### Description

When creating a Micronaut application that uses Java 21 

```sh
mn create-app --build=gradle_kotlin --jdk=21 --lang=java --test=junit com.example.demo
```

and the run `./gradlew dockerfile` or `./gradlew dockerfileNative`, Micronaut generates Dockerfile based on a Java 17 runtime. The native image cannot be built and the JAR based Docker image fails to start.

#### DockerfileNative

```Dockerfile
FROM ghcr.io/graalvm/native-image-community:17-ol9 AS graalvm
...
```
#### Dockerfile

```Dockerfile
FROM eclipse-temurin:17-jre-focal
...
```

#### Workaround

As a workaround for the JAR based Dockerfile I can define a different base image such as

```kotlin
tasks.named<MicronautDockerfile>("dockerfile") {
    baseImage = "azul/zulu-openjdk:21-jre"
}
```

For the DockerfileNative 

```kotlin
tasks.named<io.micronaut.gradle.docker.NativeImageDockerfile>("dockerfileNative") {
    graalImage = "ghcr.io/graalvm/native-image-community:21-ol9"
}
```

### Steps To Reproduce

see actual behaviour...

### Version

4.2.3

**Comment by saw303** [comment]
relates to https://github.com/micronaut-projects/micronaut-core/issues/10046

**Comment by melix** [comment]
This is actually more a problem of Micronaut Starter which should generate the proper code for the docker file tasks (which have the `jdkVersion` property). Part of the problem with doing that automatically (or inferring, if you prefer) is that there is no standard way in Gradle to get that information consistently (we'd have to figure out if `sourceCompatibility` is used, or toolchains, or both).

**Comment by saw303** [comment]
@melix I see. I was not sure whether this is a starter or a Gradle plugin issue. Feel free to move this issue to the right repo.

**Comment by sdelamo** [comment]
@melix @saw303 what do we changes are needed in starter when selecting java 21?

**Comment by melix** [comment]
It should generate something like this:

```kotlin
tasks {
    dockerfileNative {
         jdkVersion.set("21")
    }
}
```

**Comment by timyates** [comment]
@melix  that works for `dockerfileNative`, but I cannot find the same thing for `dockerfile` 🤔

**Comment by timyates** [comment]
I will work on setting the base image if java 21 is selected

**Comment by melix** [comment]
That setting is indeed specific to our native image docker file. For the regular one we'd need to change the base image. Maybe worth creating an issue on the Gradle plugin to support it too.

**Comment by timyates** [comment]
Raised a PR to fix this in starter for 4.2.5

**Comment by saw303** [comment]
Thank you!
